File: Array/PairSum.java
package Array;

import java.util.Arrays;
import java.util.HashMap;

/**
 * PairSum Find a pair that sums to target value Using dictionary approach
 */
public class PairSum {

    public static void main(String[] args) {

        int[] arr = { 3, 7, 2, 5, 6, 3 };
        int target = 9;

        int[] pair = pairSum(arr, target);

        System.out.println(Arrays.toString(pair));

    }

    public static int[] pairSum(int[] arr, int target) {
        int pair[] = new int[2];
        HashMap<Integer, Boolean> isInArr = new HashMap<>();

        // keep elements in array into a dictionary (not counting)
        for (int i : arr) {
            isInArr.putIfAbsent(i, true);
        }

        for (int i : arr) {
            int diff = target - i;
            if (isInArr.get(diff) != null) {
                pair[0] = i;
                pair[1] = diff;
                return pair;
            }
        }

        return pair;
    }

}

File: Array/PairSum.py
'''
Given an unsorted list of integers and a target value,

find a pair that sums up to be the target value.

Sorting and then two-pointer approach
'''
# uses python's built-in list sort()
# https://wiki.python.org/moin/TimeComplexity
# python's sort uses Timsort and takes O(n log n)


def pairSum(arr, target):
    pair = []
    arr.sort()
    i, j = 0, len(arr)-1
    while i < j:
        if arr[i]+arr[j] < target:
            i += 1
        elif arr[i]+arr[j] > target:
            j -= 1
        else:
            pair.append(arr[i])
            pair.append(arr[j])
            return pair

    return pair


def main():
    arr = [3, 7, 2, 5, 6]
    target = 9

    pair = pairSum(arr, target)

    print(pair)


if __name__ == "__main__":
    main()
